File: internal/lua/mapper/ws_msg.go
package mapper

import (
	lua "github.com/yuin/gopher-lua"
	"gitlab.com/marsskom/burro/internal/model"
)

func WSMsgToLua(L *lua.LState, msg *model.WSMessage) (*lua.LTable, error) {
	tbl := L.NewTable()

	L.SetField(tbl, "direction", lua.LString(msg.Direction))
	L.SetField(tbl, "opcode", lua.LNumber(msg.OpCode))
	L.SetField(tbl, "opcode_name", lua.LString(opcodeName(msg.OpCode)))
	L.SetField(tbl, "timestamp", lua.LNumber(msg.Timestamp))

	if len(msg.Data) == 0 {
		L.SetField(tbl, "data", lua.LNil)
	} else {
		L.SetField(tbl, "data", lua.LString(msg.Data))
	}

	if msg.Text == "" {
		L.SetField(tbl, "text", lua.LNil)
	} else {
		L.SetField(tbl, "text", lua.LString(msg.Text))
	}

	return tbl, nil
}

func opcodeName(op model.WSOpCode) string {
	switch op {
	case model.WSText:
		return "text"
	case model.WSBinary:
		return "binary"
	case model.WSClose:
		return "close"
	case model.WSPing:
		return "ping"
	case model.WSPong:
		return "pong"
	case model.WSContinuation:
		return "continuation"
	default:
		return "unknown"
	}
}

File: runtime/plugins/luaplugin/scripts/metric/ws_msg.lua
---Exists only in `ws_msg.lua`
---@alias WSDirection
---| "client_to_server"
---| "server_to_client"

---@alias WSOpCode
---| 0   -- continuation
---| 1   -- text
---| 2   -- binary
---| 8   -- close
---| 9   -- ping
---| 10  -- pong

---@class WSMessage
---@field direction WSDirection
---@field opcode WSOpCode
---@field opcode_name string
---@field data string?
---@field text string?
---@field timestamp integer

---@type WSMessage
ws = ws or {}

local value, stop = shared.load_kv_json("ws_msg")
if stop or value == nil then
	return
end

if value[ws.opcode_name] == nil then
	value[ws.opcode_name] = 0
end

value[ws.opcode_name] = tonumber(value[ws.opcode_name]) + 1

stop = shared.save_kv_json(value, "ws_msg")
if stop then
	return
end
